packages/dashboard/src/app/views/myStories/filters/taxonomy/useTaxonomyFilters.js

```diff
@@ -0,0 +1,154 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import {
+  useCallback,
+  useEffect,
+  useState,
+  useMemo,
+  useRef,
+} from '@googleforcreators/react';
+
+/**
+ * Internal dependencies
+ */
+import useApi from '../../../../api/useApi';
+
+/**
+ * Hook used for taxonomy filters logic.
+ * Initializes the taxonomy filters data.
+ *
+ * @return {Array} [taxonomies, queryTaxonomyTerms] initial taxonomy filters data and a function to query the terms.
+ */
+
+function useTaxonomyFilters() {
+  const [taxonomies, setTaxonomies] = useState([]);
+  const termPromises = useRef([]);
+
+  const { getTaxonomies, getTaxonomyTerms } = useApi(
+    ({
+      actions: {
+        taxonomyApi: { getTaxonomies, getTaxonomyTerms },
+      },
+    }) => ({
+      getTaxonomies,
+      getTaxonomyTerms,
+    })
+  );
+
+  /**
+   * Maps the fetched taxonomy data back to parent taxonomy.
+   *
+   * @return {void}
+   */
+  const _setTaxonomiesTermData = useCallback(async () => {
+    const fetched = await Promise.all(termPromises.current);
+    const taxonomyObject = {};
+    for (const arr of fetched) {
+      // grab the first elements 'taxonomy' which should map to the parents 'slug'
+      const key = arr.at(0)?.taxonomy;
+      if (key) {
+        taxonomyObject[key] = arr;
+      }
+    }
+    setTaxonomies((current) => {
+      return current.map((c) => {
+        const data = taxonomyObject[c.slug] || [];
+        return {
+          ...c,
+          data,
+        };
+      });
+    });
+  }, [setTaxonomies]);
+
+  /**
+   * Query individual taxonomy data.
+   * This is needed to initialize the primaryOptions and needed to search and set queriedOptions.
+   *
+   * @param {Object} taxonomy object holding taxonomy data, including restBase and restPath used to fetch individual taxonomy terms
+   * @param {string} search string use to query taxonomy by name
+   * @return {Array} taxonomy terms
+   */
+  const queryTaxonomyTerms = useCallback(
+    async (taxonomy, search) => {
+      const { restBase, restPath } = taxonomy;
+      const terms = await getTaxonomyTerms(restPath, { search, per_page: -1 });
+      return terms.map((t) => ({
+        ...t,
+        restBase,
+        restPath,
+      }));
+    },
+    [getTaxonomyTerms]
+  );
+
+  /**
+   * Query all the taxonomies.
+   * This should only be needed once get all the taxonomies and set individual term data.
+   *
+   * @see queryTaxonomyTerms
+   */
+  const queryTaxonomies = useCallback(async () => {
```

Comment:
The whole logic with the termPromises ref etc. is a bit hard to read/follow. I suggest creating a follow-up ticket to improve readability a bit.

Reply:
🤔 I agree, the logic is a bit hard to read, I see a way to make a small change that'll simplify it 😅